### PR TITLE
Patch issue #6035

### PR DESCRIPTION
**Pull request by afvincent**
Here is a proposition of patch that should fix issue #6035, without breaking `annotate` (at least the relevant unit test seems still OK). 

As @efiring suggested, I tried not to directly modify the value of the attribute `mutation_scale` (as it is exposed as an argument). Basically, the idea is to scale `mutation_scale` accordingly to the DPI correction (`dpi_cor`), inside `get_path_in_displaycoord`, which is called in the `draw` method of the `FancyPatchArrow` class. This way, the head (and tail) size of a `FancyPatchArrow` should not depend on the DPI value. 

Besides, I removed the same kind of scaling operation that was done “externally” in `annotate` (in `text.py`), since it is now performed directly inside `FancyArrowPatch`. As `test_fancyarrow` still is OK, I would assume nothing has been broken.

I also added an `image_comparison` test (well in fact two, one for the PNG export at 100 DPI and another one at 200 DPI). Visually both exported pictures look the same in term of the size of the arrow head. Concerning the code in itself, it seems a bit clunky to me to use two separate test functions only to export with two different DPI values identical figures, but I did not managed to get a working `class Test_*`… (Nevertheless, the current situation should do the job.)

PS: I struggled a bit with git (I “had to” remove one branch from my fork): I hope everything is fine with the commits for the PR.

### CODE REVIEW

**Comment by efiring** [comment]
@afvincent, this is encouraging.  In working through this, have you figured out what "mutation_scale" and "dpi_cor" really *are*?  Can they be explained in the documentation so that a user will be able to understand how to use them?  It looks to me like `dpi_cor` should not be exposed at all--it should be a private internal variable, not a kwarg.  It's not used in any of our examples. I'm not saying this is a change that can or should be made in this PR, but maybe we can lay out a strategy for improving the API and documentation.

**Comment by afvincent** [comment]
## 0 - TL;DR
Somebody corrects me if I am wrong but I would say that:
- `dpi_cor` is a convenient scaling factor, **used only internally** to convert between points and pixels and shouldn't be exposed to the user;
- `mutation_scale` is a reference scale **in points**, used to compute the size of elements like the head or the tail of an arrow.
 
For people who have time, here are some explanations.

## 1 - What is dpi_cor?
Well `grep dpi_cor patches.py` shows that `dpi_cor` is only used to do things like:
`self.set_dpi_cor(renderer.points_to_pixels(1.))` or `dpi_cor = renderer.points_to_pixels(1)`, 
and then: 
`X = X * dpi_cor` or `self.set_X(X * dpi_cor)`, 
with X assumed to be in points (for example X is `mutation_scale` or `linewidth`). 

**So I would say that `dpi_cor` is simply a scaling factor, used to convert the relevant values from points to pixels when drawing the figure.** I am not convinced it is very wise to expose it to the user as it is currently done in `FancyArrowPatch` or `ConnectionPatch`: the provided value will be overwritten anyway.

## 2 - What is mutation_scale?
Concerning `mutation_scale`, it is less clear to me. I would guess it was firstly introduced with the `BoxStyle` and `FancyBboxPatch` classes. Indeed their docstrings explain:
> An instance of any boxstyle class is an callable object, whose call signature is:
> `__call__(self, x0, y0, width, height, mutation_size, aspect_ratio=1.)`
> and returns a :class:`Path` instance. *x0*, *y0*, *width* and *height* specify the location and size of the box to be drawn. 
> *mutation_scale* determines the overall size of the mutation (by which I mean the transformation of the rectangle to the fancy box).  

NB: from reading other method signatures, I would guess that  `mutation_scale` is the same thing as `mutation_size` (maybe it would be worth to thoroughly check it and change it for more consistent notations among the complete `patches.py`?)

Then `ArrowStyle` is defined, mainly for `FancyArrowPatch` and one can also read in its docstring:
> An instance of any arrow style class is a callable object, whose call signature is:
> `__call__(self, path, mutation_size, linewidth, aspect_ratio=1.)`
> and it returns a tuple of a :class:`Path` instance and a boolean value. *path* is a :class:`Path` instance along which the arrow will be drawn. 
> *** *mutation_size* and *aspect_ratio* have the same meaning as in :class:`BoxStyle`. ***

If I understood correctly everything, it is then for example called from inside `FancyArrowPatch`, when doing:
```python
_path, fillable = self.get_arrowstyle()(_path,
                                        self.get_mutation_scale() * dpi_cor,
                                        self.get_linewidth() * dpi_cor,
                                        self.get_mutation_aspect()
                                       )
# NB: the 2nd line tends to support the idea that mutation_size
# is nothing else than mutation_scale, doesn't it?
```

`FancyArrowPatch` docstring explains also:
> *mutation_scale* : a value with which attributes of arrowstyle (e.g., head_length) will be scaled. default=1.

**Note that no units are precised.** And this description is confirmed by this kind of lines in some methods of `ArrowStyle`:
```python
head_length = self.head_length * mutation_size
```
and one can also read:
> *head_length* and *head_width* determines the size of the arrow relative to the *mutation scale*.

So I would say that both head_length and head_width are unitless, while **`mutation_scale` has a unit**, as we will just just after.

Further, in `ConnectionPatch`, is it written:
> mutation_scale:   default is text size (in points).

**So `mutation_scale` does apparently have a unit (the point)**.  (And here its **default is not 1 anymore**, but _text_size_, which by the way is arbitrary fixed to 10 (points) if one reads the `__init__`.)

## The case of the class Annotation
`FancyArrowPatch` is heavily used in `Annotation` (located in `text.py`), used for example by `plt.annotate` if I am correct. Here it is also written
> mutation_scale: default is text size (in points)

as in `ConnectionPatch`. However this time, `mutation_scale` is indeed sized as the text, through lines like:
```python
ms = d.pop("mutation_scale", self.get_size())
self.arrow_patch.set_mutation_scale(ms)
```
that pass argument to the ad-hoc `FancyArrowPatch` instance.

## 3  - Tentative of conclusion about mutation_scale
So from what I read and understood, I see `mutation_scale` as a reference scale, **expressed in points**, that is mainly used to compute the dimensions of the head and tail of the arrow when drawing the figure. I will try to see if one can make the docstring a bit clearer about it.

_Small remark_: from my point of view, the fact that in raw `FancyArrowPatch` the default value is 1 (pt ?), while in derived classes like `ConnectionPatch` or `Annotation` it is much bigger (10 pt or fontsize), may be a bit disturbing for users (at least it is for me ^^). Especially when plotting arrow-only annotations: I get significantly different behaviors (when relying on default values), eventhough I would expect similar results (but maybe it is simply because I consider Annotation as ~ FancyArrow + Text).

**Comment by afvincent** [comment]
About the commit _1df0f7e_

### Failures in test_axes
I think it solves some of the problems that I didn't expected when patching `FancyArrowPatch` last time. It should mainly correct the fact that `test_polar_annotations` and `test_polar_coord_annotations` were failing. Well they still do, but I am rather convinced it is now essentially due to a problem of tolerance on these two unit tests: reported RMS are rather low, something like 0.003 and when I look at the `*-failed-diff.png`, it seems uniformly black to me. See the example:
example_of_failed-diff
Or am I missing something? By the way, the arrow now behaves correctly :) (it had a different size before with the previous commit).

### Failures in test_streamplot
Additionally, I finished to convinced myself that the unit tests `test_colormap`, `test_linewidth` and `test_masks_and_nans` (in `test_streamplot`) are failing more or less because of the former wrong dpi-dependent behavior of `FancyArrowPatch`. It is rather clear on the following picture that the exported streamplot depends on the DPI value with the vanilla version of `FancyPatchArrow` (right panels) while it does not with the patched version (left panels): 
looking_for_streamplot_failure_origin

Currently the problem is that the two upper panels (@ 100 DPI) are not perfectly identical: the arrow sizes are not the same. It is confirmed by the `*-failed-diff.png` too. Eventhough the former reference behavior might not be that good anyway (given its results for a higher value of DPI), I will investigate why the arrow size changes with the new `FancyArrowPatch`.

### Failure in test_transforms
From the `*-failed-diff.png`, the problem with `test_pre_transform_plotting` is related to a portion of the image with a streamplot, so same “problem” as before.

**Comment by tacaswell** [comment]
I do not have time to say anything technically meaningful, but thank you for your detailed work on this!

```pytb
======================================================================
ERROR: matplotlib.tests.test_artist.test_clipping.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python3.5/site-packages/nose/case.py", line 198, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 249, in do_test
    self._tol, in_decorator=True)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 332, in compare_images
    rms = calculate_rms(expectedImage, actualImage)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 246, in calculate_rms
    abs_diff_image = abs(expectedImage - actualImage)
ValueError: operands could not be broadcast together with shapes (540,720,3) (270,360,3) 
```

Those failures worry me a bit.

I suspect that if you track down the issue with the stream plots, the polar plot issue will go away.

The new stream plot does look a bit better to my eyes, it would not take much to convince me that the test images should be updated.

**Comment by afvincent** [comment]
It is weird, I have no problem with `test_artist` on my machine (CentOS 7, Python 2.7 in a conda env, with Matplotlib 1.5.1 + my last commit):
 ```python
In [3]: run test_artist.py
/home/adrien/matplotlib-git/matplotlib/lib/matplotlib/__init__.py:885: 
UserWarning: axes.color_cycle is deprecated and replaced with axes.prop_cycle; 
please use the latter.  # I still didn't have a look into the new color_cycle thing, 
                        # but that should irrelevant for our concern, shouldn't it? 
warnings.warn(self.msg_depr % (key, alt_key))
/home/adrien/matplotlib-git/matplotlib/lib/matplotlib/testing/decorators.py:363: 
UserWarning: test module run as script. guessing baseline image locations 
warnings.warn('test module run as script. guessing baseline image locations')
.........
----------------------------------------------------------------------
Ran 9 tests in 2.641s

OK
```

**Comment by QuLogic** [comment]
Travis is running this PR against latest master, not 1.5.1. There have been many changes since then, but the most likely one is a change in the default DPI.

**Comment by afvincent** [comment]
Just to say that I didn't have time lately to dedicate checking the final details of this PR, but I am still working on it to understand why some unit tests fail, especially the ones that do by only a few pixels.

**Comment by afvincent** [comment]
### About the failures of `test_axes`

Both `test_polar_annotations` and `test_polar_coord_annotations` in `test_axes` are still failing. I made manual comparisons (with `scipy.misc.imread` and Co…) and the baseline images and the result images are 100% identical for the SVG and PDF versions (`*_svg.png` and `*_pdf.png`). The problem seems to only occur with the PNG export, where 4 and 12 pixels are indicated as not identical between baseline and result. However, I am not sure it is really a problem. Here is a picture where I highlighted the region with the-- only 4 in this case-- problematic pixels (the zoom factor is x4):
analyse_of_differences_between_polar_axes

Without ImageMagick, one can hardly notice the difference of the 4 pixels… (Here are the 3 inset pictures for those who would like to compare by themselve: [ROI_comparisons.tar.gz](https://github.com/matplotlib/matplotlib/files/168250/ROI_comparisons.tar.gz))

I am wondering how relevant are these two failures. *Are such tiny differences something usual ? Plus the fact it only affects the PNG export, and not the SVG and the PDF ones ?*

### About the other failures

I am still investigating the case of the streamplot failures, but I am rather confident it is only due to the former dpi-dependent behavior. Besides, I expect that solving that will also solve the problem with `test_transforms` (that seems very similar).

**Comment by jenshnielsen** [comment]
Thanks for the detailed investigation. I tend to agree with you that this change is not significant and we can just replace the image with the new version. It's really hard to set the tolerance for image comparison tests. We used to have a much higher tolerance but that resulted in a number of issues going unnoticed over the years.

**Comment by tacaswell** [comment]
@afvincent Echoing Jens, thank you for the very detailed work tracking this down!  For changes that small please replace the images.

File: lib/matplotlib/patches.py [hunk 1]
```diff
@@ -4057,7 +4057,7 @@ def __init__(self, posA=None, posB=None,
     def set_dpi_cor(self, dpi_cor):
         """
         dpi_cor is currently used for linewidth-related things and
-        shrink factor. Mutation scale is not affected by this.
```
Comment (tacaswell) [inline lib/matplotlib/patches.py]: just drop the 'not', in the near future no one will remember the old behavior and the `'now'` will just add confusion.

File: lib/matplotlib/text.py [hunk 2]
```diff
@@ -2159,11 +2158,17 @@ def _update_position_xytext(self, renderer, xy_pixel):
                         " use 'headlength' to set the head length in points.")
                 headlength = d.pop('headlength', 12)
 
-                to_style = self.figure.dpi / (72 * ms)
-
-                stylekw = dict(head_length=headlength * to_style,
-                               head_width=headwidth * to_style,
-                               tail_width=width * to_style)
+                # Old way:
```
Comment (tacaswell) [inline lib/matplotlib/text.py]: please remove this comment block

If anyone has to do forensics on this part of the code recursive git-blame is their friend.   I think going forward the discussion of old and new way will just lead to confusion

**Comment by tacaswell** [comment]
:+1: I am sold on this once the tests are updated.

File: lib/matplotlib/patches.py [hunk 3]
```diff
@@ -4223,10 +4223,10 @@ def get_path_in_displaycoord(self):
             _path = self.get_transform().transform_path(self._path_original)
 
         _path, fillable = self.get_arrowstyle()(_path,
-                                                self.get_mutation_scale(),
-                                                self.get_linewidth() * dpi_cor,
-                                                self.get_mutation_aspect()
-                                               )
+                                        self.get_mutation_scale() * dpi_cor,
```
Comment (QuLogic) [inline lib/matplotlib/patches.py]: Does this actually pass pep8?

File: lib/matplotlib/tests/test_arrow_patches.py [hunk 4]
```diff
@@ -5,7 +5,7 @@
 
 import matplotlib.pyplot as plt
 from matplotlib.testing.decorators import image_comparison
-import matplotlib
+import matplotlib.patches as mpatches  # adopting scipy import conventions
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_arrow_patches.py]: Remove this comment; it's matplotlib's convention too.

File: lib/matplotlib/tests/test_arrow_patches.py [hunk 5]
```diff
@@ -52,6 +52,51 @@ def test_boxarrow():
                   bbox=dict(boxstyle=stylename, fc="w", ec="k"))
 
 
+def prepare_fancyarrow_dpi_cor_test():
+    """
+    Convenience function that prepares and return a FancyArrowPatch. It aims
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_arrow_patches.py]: "prepares and return" -- the two verbs do not agree in tense.

File: lib/matplotlib/tests/test_arrow_patches.py [hunk 6]
```diff
@@ -52,6 +52,51 @@ def test_boxarrow():
...
+    return fig2
+
+
+@image_comparison(baseline_images=['fancyarrow_dpi_cor_100dpi'],
+                  remove_text=True, extensions=['png'],
+                  savefig_kwarg=dict(dpi=100))
+def test_fancyarrow_dpi_cor_100dpi():
+    """
+    Check the export of a FancyArrowPatch @ 100 DPI. FancyArrowPatch is
+    instanciated through a dedicated function because another similar test
```
Comment (QuLogic) [inline lib/matplotlib/tests/test_arrow_patches.py]: instantiated

**Comment by afvincent** [comment]
@QuLogic and @tacaswell : Thank you for the corrections and typos. I hope everything is fine now.

_Note to myself: still have to generate and to update the test images._

**Comment by afvincent** [comment]
I will update the test images tomorrow. However, I still have some questions about the failures in `test_streamplot` and `test_transforms`.

## Failures in test_streamplot
Definitively, before the patch the relative size of the arrows was different between the PNG image and the PDF/SVG images. With the patch, the tests are OK with PDF/SVG images, but they fail with the PNG images. Looking at the `failed-diff` files shows that it is due to the fact that the arrows have become bigger than previously (but at least they are now of the same relative size as in the PDF and SVG images). 

I played a bit with the argument `arrowsize` in `streamplot`, and it appears that `arrowsize = 0.72` (default value is 1) leads to successful PNG image tests. But of course the PDF/SVG image tests fail in that case. As the parameter `arrowsize`is only used in `streamplot` to compute an *internal* kwarg (`mutation_scale = 10 * arrowsize`), the old PNG behavior could be transparently recovered for the user (with `mutation_scale = 7.2 * arrowsize`).

__So which one of the two possible behaviors does one want to choose as the correct one:__
* A/ the one of the PDF/SVG images (and the PNG baseline images have to be updated, to have “big” arrows too) ?
* B/ the one of the former PNG images (and the PDF/SVG baseline images have to be updated, as well as the internal `mutation_scale` computation must be modified in `streamplot`, to keep “small” arrows)?

## Failures in test_transforms
With the patch and without playing with `arrowsize` in `streamplot`, only the PNG image test fails, because the arrow size is bigger than before (as for `test_streamplot`)

Adding `arrowsize = 0.72` in the `streamplot` call of the test also gives the same arrow size as before in the PNG images. However, a clipping issue seems to appear in this case: a small blank space on the right disappears and everything in the picture is thus moved a bit toward the right edge; all the tests then fail (PNG/PDF/SVG).

## Personal conclusion
I would suggest solution A/, i.e. updating the PNG images for `test_streamplot` and `test_transform`, as i) it does not requires changing anything else and ii) it avoids the clipping issue in the `test_transforms` when playing with the argument `arrowsize`.

**Comment by tacaswell** [comment]
:+1: on A then.

**Comment by afvincent** [comment]
Something puzzles me with Git. As I am a newbie with Git, I simply followed the instructions on [http://matplotlib.org/devel/gitwash/development_workflow.html](url) to stay up-to-date with upstream:
```
git checkout master
git fetch upstream
git merge --ff-only upstream/master
```
However, it appears that the files in my branch (`patch-issue-6035`) are not impacted by this. *Am I supposed to merge or “rebase” something (in addition to the 3 previous commands) to see the modifications that have been accepted and merged in the upstream master branch?*

For example, the names of the `streamplot` test images have been changed since I branched, and one additional image test has been added to `test_streamplot`  (#6011 ). Unfortunately, its PNG test image will suffer from the very same problem as the other ones do. So I generated the updated PNG image `streamplot_startpoints.png` (with the new name convention, btw):
streamplot_startpoints
but I don't know how to commit it properly… Besides, for the same reason, the names of the other PNGs I've just committed for `test_streamplot` follow the old convention.

**Comment by tacaswell** [comment]
That your branch did not change is a feature, not a bug . 

This is a long read http://tom.preston-werner.com/2009/05/19/the-git-parable.html but is a great description of how git works.

Branches are pointers into a tree of commits, each of which only knows it parent(s).  Moving the master branch to match upstream master should not affect your feature branch.  To rebase, see https://github.com/matplotlib/matplotlib/pull/2742/files (which I really should finish cleaning up the biksheding over 'upstream' vs 'matplotlib' as a remote name and get merged).

The TL;DR is 

```shell
git checkout your_branch
git rebase matplotlib/master
```

and then follow the on-screen instructions.  If you get too stuck let me know.

**Comment by afvincent** [comment]
@tacaswell : thank you for the documents (but I think I'll have to read again some of them before being comfortable with Git). Your doc on rebase operation was very useful, and I hope it will be published soon :).

I rebased my branch with
```
git branch tmp-6035 patch-issue-6035 # just in case, as recommended in numpy's doc
git checkout your_branch
git rebase upstream/master
```
and I decided to drop the changes to the PNGs that I did yesterday, in favor of the the ones made in #6011 . I was planning to then re-commit my updated PNGs (+ the new one on the start_points test), all with correct new names, after having checked that all the other tests were OK. 

Unfortunately, it seems that most of the PDF image tests are now failing. It is particularly weird in `test_axes`, where tests about things I have not modified at all (for example boxplots…) are failing. Looking at the `*-failed-diff.png` files, it looks like it is due before all to the fonts of the ticklabels, that are slightly different from before (of a few pixels); I don't know why…

**Comment by tacaswell** [comment]
Are you using the locally built freetype?  There can be slight variation in the font rendering due to changes in freetype.  To get to zero-tolerance we have pinned to a specific freetype version (2.6.1) for testing.

I suggest pushing up your rebased branch (you will need to force push) to see what travis has to say before getting to worried about the pdf failures.

It may also be neccessary to do a `git clean -xfd` as there have been changes to the c parts of the code and distutls is not always clever enough to tell when those need to be rebuilt.

**Comment by afvincent** [comment]
I force pushed the rebased branch as suggested. Normally, the PNG image tests that were failing before (in `test_streamplot`, `test_axes` and `test_transforms`) should be OK now (they are on my computer). Having to re-update the one in `test_transforms` surprised me a bit, but it was a difference of only 2 pixels so I didn't really looked into it and simply updated the PNG (if it was a mistake, just tell me and I will give it more attention).

@tacaswell  Fingers crossed about the PDF image tests and freetype: if the development matplotlib on my computer is using the OS freetype, it is indeed the version 2.6.3 and not the 2.6.1 one… 

Just for my knowledge: `git clean -xfdn` shows me that it would delete more or less all the .pyc and .so files. So the idea is to force the creation of fresh ones, which might solve the issue with the faulty PDF image tests?

**Comment by tacaswell** [comment]
@afvincent This needs a rebase again, not sure what we changed under you.

**Comment by afvincent** [comment]
@tacaswell  Is it very urgent or can it wait for a week? Because I shouldn't have access to a correct Internet connection (or even a computer...) before next weekend. (However, if it is really urgent because of some milestone for example, I'll try to find what's needed to rebase before).

**Comment by tacaswell** [comment]
I think it can wait a week.  If it can not, would you be offended if someone else did the rebase (+ sorting out the clipping test failures)?

**Comment by afvincent** [comment]
Of course I have no problem with someone else doing the rebase if needed! (And if it is still open when I will have something else than a smartphone as Internet access, I will take care of it.)

**Comment by afvincent** [comment]
@tacaswell I did another rebase, skipping the failing 3 last patches (2110e5e, 56f337f and 83fb8d1) about failing PNG baseline test images, and then re-updated these test images on my local repository. Unfortunately, `git push origin patch-issue-6035` gives me the following error message (I apologize, my OS locale is French):
```
 ! [rejected]        patch-issue-6035 -> patch-issue-6035 (non-fast-forward)
error: impossible de pousser des références vers 'git@github.com:afvincent/matplotlib.git'
astuce: Les mises à jour ont été rejetées car la pointe de la branche courante est derrière
astuce: son homologue distant.
```
Homemade translation:
```
 ! [rejected]        patch-issue-6035 -> patch-issue-6035 (non-fast-forward)
error: impossible to push references toward 'git@github.com:afvincent/matplotlib.git'
tip: updates were rejected because current branch head is behind the distant head
```
So I guess I managed once more to get git in an unwanted situation… If it matters, just before rebasing, I used 
```
git checkout master
git fetch upstream
git merge --ff-only upstream/master
``` 

Wouldn't it be easier to create a new branch from the current master, on which I will report the changes of the current patch, and then to open a new PR?


PS : I also did ` git clean -xfdn`, but then I had issues with some tests scripts, failing when trying to load `matplotlib._path` for example. Re-running `python setup.py develop` in my python virtual environment made it work again. However I still have a lot of failing tests, due to what seems to be tiny differences of fonts (but these failures don't seems to occur on Travis & Co)

**Comment by tacaswell** [comment]
You need to do

git push --force origin patch-issue-6035

To tell guy to do something that might lose commits.

**Comment by afvincent** [comment]
The previous updated PNGs test images seemed to be still failing due to difference in the fonts of the axis labels, so I just re-commited the same images but produced on a computer than seemed to be free of the font issues I have on my personal computer. Apparently the associated failures have disappeared after this last commit.

So now only the failures of `test_artist.test_clipping` still remain...

**Comment by afvincent** [comment]
I am a bit lost with the failures of `test_clipping`… After a look at the code of this test, I do not see any method or other stuff I might have directly modified, but yet Travis complains:
```python
======================================================================
FAIL: matplotlib.tests.test_artist.test_clipping.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 255, in do_test
    self._tol, in_decorator=True)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 337, in compare_images
    rms = calculate_rms(expectedImage, actualImage)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 248, in calculate_rms
    "actual size {1}".format(expectedImage.shape, actualImage.shape))
ImageComparisonFailure: image sizes do not match expected size: (600, 800, 3) actual size (300, 400, 3)

======================================================================
FAIL: matplotlib.tests.test_artist.test_clipping.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 255, in do_test
    self._tol, in_decorator=True)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 337, in compare_images
    rms = calculate_rms(expectedImage, actualImage)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 248, in calculate_rms
    "actual size {1}".format(expectedImage.shape, actualImage.shape))
ImageComparisonFailure: image sizes do not match expected size: (432, 576, 3) actual size (216, 288, 3)

======================================================================
FAIL: matplotlib.tests.test_artist.test_clipping.test
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/travis/build/matplotlib/matplotlib/venv/lib/python2.7/site-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 55, in failer
    result = f(*args, **kwargs)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/decorators.py", line 255, in do_test
    self._tol, in_decorator=True)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 337, in compare_images
    rms = calculate_rms(expectedImage, actualImage)
  File "/home/travis/build/matplotlib/matplotlib/lib/matplotlib/testing/compare.py", line 248, in calculate_rms
    "actual size {1}".format(expectedImage.shape, actualImage.shape))
ImageComparisonFailure: image sizes do not match expected size: (540, 720, 3) actual size (270, 360, 3)
```

Looks like (at least) one of the problems is that the DPI value use when saving the images is half the expected one. QuLogic suggested a while ago it might be related to a change in the default DPI value, but then should'nt there be a lot more of test failures (basically all the image comparison tests)? Furthermore, isn't it a bit strange that even the PNG conversions of the PDF and SVG images are also saved with half the expected resolution?

**Comment by afvincent** [comment]
@jenshnielsen @tacaswell 

On a different topic than the remaining issues with `test_artist.test_clipping` I am still struggling with: `axis_grid` from `mpl_toolkits` uses `FancyArrowPatch` for some inheritance (at least for setting an arrow axis style as in [demo_axisline_style](http://matplotlib.org/examples/axes_grid/demo_axisline_style.html)). I tested rapidly : indeed the size of the arrow heads of the axes was dpi-dependent before the patch, which is corrected after it. This change was unnoticed before because none of the mpl_toolkits tests seems to make use of such an “arrow axis”. *How is this a problem? And should a dedicated test be written?*

**Comment by tacaswell** [comment]
If you feel compelled to write a test we will not protest, but a vast majority of mpl_toolkit is not tested and cleaning that up is not a burden we want to put on this PR.  It is also not very well documented which means that  it's usage is probably pretty low.

It also sounds like you have made things better not worse?

**Comment by afvincent** [comment]
I would say the situation is better now, as the size of the arrow heads of the axes are not dpi-dependent anymore. Before the patch, `demo_axisline_style` produced axes with barely visible arrow head when saved with 300 DPI… But as with the rest of this PR, the arrows produced by default for the axes are now a bit bigger than before (as for example as with `test_streamplot`).

**Comment by jenshnielsen** [comment]
The issue with `test_artist.test_clipping` is caused by it executing right after your new tests. 

For reasons unknown `prepare_fancyarrow_dpi_cor_test` is getting executed as a test even if it's not called `test_` and since it's not marked as `@image_comparison` or `@cleanup` the style (mainly change to dpi I think) is leaked to the next test. 

Making the method private by renaming it to `__prepare_fancyarrow_dpi_cor_test` seems to fix the issue.

**Comment by QuLogic** [comment]
Replaced by #6504.
